controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -733,24 +878,44 @@ func (pp *portPublisher) endpointSliceToAddresses(es *discovery.EndpointSlice) A
 					pp.log.Errorf("Unable to create new address:%v", err)
 					continue
 				}
-
+				ppID := podPortID{
+					name:      endpoint.TargetRef.Name,
+					namespace: endpoint.TargetRef.Namespace,
+					port:      resolvedPort,
+				}
+				if _, ok := serverOpaquePodPorts[ppID]; ok {
+					id := PodID{
+						Name:      endpoint.TargetRef.Name,
+						Namespace: endpoint.TargetRef.Namespace,
+					}
+					if ports, ok := opaquePodPorts[id]; ok {
+						ports[resolvedPort] = struct{}{}
+					} else {
+						ports := map[Port]struct{}{resolvedPort: {}}
+						opaquePodPorts[id] = ports
+					}
+				}
 				if endpoint.Hints != nil {
 					zones := make([]discovery.ForZone, len(endpoint.Hints.ForZones))
 					copy(zones, endpoint.Hints.ForZones)
 					address.ForZones = zones
 				}
-
-				addressSet.Addresses[id] = address
+				addresses[id] = address
 			}
 		}
 
 	}
-
-	return addressSet
+	return AddressSet{
+		Addresses:      addresses,
+		Labels:         metricLabels(es),
+		OpaquePodPorts: opaquePodPorts,
+	}
 }
 
 func (pp *portPublisher) endpointsToAddresses(endpoints *corev1.Endpoints) AddressSet {
+	serverOpaquePodPorts := pp.getServerOpaquePodPorts()
```

Comment:
Rather than querying for all servers AND all pods selected by those servers, I think we can query for all servers, and then iterate through the servers and the endpoints in the update.  If the endpoint is a pod and the server selects that pod, and the endpoint's port matches the server's port, then consider the address to be opaqueProtocol.

Practically speaking, I think that means getting all servers here and then passing that server list into newPodRefAddress.  The newPodRefAddress function should take the server list into account and if any of the servers match the resolvedPort and the selector matches the pod labels, then we can set opaqueProtocol on the address.